sdk/keyvault/keyvault-certificates/tests/utils/recorder.ts

```diff
@@ -0,0 +1,434 @@
+      .replace(/</g, "lt")
+      .replace(/>/g, "gt")
+      .replace(/=/g, "eq")
+      .replace(/\W/g, "");
+  }
+
+  /**
+   * Additional layer of security to avoid unintended/accidental occurrences of secrets in the recordings
+   * */
+  protected filterSecrets(recording: string): string {
+    let updatedRecording = recording;
+    for (const k of Object.keys(replaceableVariables)) {
+      const escaped = escapeRegExp(env[k]);
+      updatedRecording = updatedRecording.replace(
+        new RegExp(escaped, "g"),
+        replaceableVariables[k]
+      );
+    }
+    for (const map of replacements) {
+      updatedRecording = map(updatedRecording);
+    }
+    return updatedRecording;
+  }
+
+  public abstract record(): void;
+  public abstract playback(): void;
+  public abstract stop(): void;
+}
+
+class NockRecorder extends Recorder {
+  constructor(testHierarchy: string, testTitle: string) {
+    super("node", testHierarchy, testTitle, "js");
+  }
+
+  public record(): void {
+    nock.recorder.rec({
+      dont_print: true
+    });
+  }
+
+  public playback(): void {
+    // This path makes sense when tests are called through dist-test/index.node.js
+    // If tests are called directly, this would need to be `../../recordings/`.
+    const path = "../recordings/" + this.filepath;
+    this.uniqueTestInfo = require(path).testInfo;
+  }
+
+  public stop(): void {
+    const importNock = "let nock = require('nock');\n";
+    const fixtures = nock.recorder.play();
+
+    // Create the directories recursively incase they don't exist
+    try {
+      // Stripping away the filename from the filepath and retaining the directory structure
+      fs.ensureDirSync(
+        "./recordings/" + this.filepath.substring(0, this.filepath.lastIndexOf("/") + 1)
+      );
+    } catch (err) {
+      if (err.code !== "EEXIST") throw err;
+    }
+
+    const file = fs.createWriteStream("./recordings/" + this.filepath, {
+      flags: "w"
+    });
+
+    // Some tests expect errors to happen and, if a writing error is thrown in one of these tests, it may be captured in a catch block by accident,
+    // resulting in unexpected behavior. For this reason we're printing it to the console as well
+    file.on("error", (err: any) => {
+      console.log(err);
+      throw err;
+    });
+
+    file.write(
+      importNock + "\n" + "module.exports.testInfo = " + JSON.stringify(this.uniqueTestInfo) + "\n"
+    );
+
+    for (const fixture of fixtures) {
+      // We're not matching query string parameters because they may contain sensitive information, and Nock does not allow us to customize it easily
+      const updatedFixture = fixture.replace(/\.query\(.*\)/, ".query(true)");
+      file.write(this.filterSecrets(updatedFixture) + "\n");
+    }
+
+    file.end();
+
+    nock.recorder.clear();
+    nock.restore();
+  }
+}
+
+class NiseRecorder extends Recorder {
+  private readonly sasQueryParameters = ["se", "sig", "sp", "spr", "srt", "ss", "st", "sv"];
```

Comment:
What's this for?

Reply:
This is taken from Harsha's code! I will ask him.